### DISCUSSION TITLE
Lock Screen Notification Text Color Inverted in Dark Mode

### DISCUSSION THREAD

**Discussion by nickb24** [NightscoutFoundation/xDrip#3617] [Bug reports]
Running on a Pixel 7 Pro with this version:
![3](https://github.com/user-attachments/assets/a4a165cf-b3ce-4869-a45f-9ea37f5fdeaa)

Running latest Nightly of xDrip - 20240807

Android OS display settings is setup to run in Dark mode.

Every time a new reading comes in with the screen off, and you turn the screen on, this is the xDrip ongoing notification seen:
![1](https://github.com/user-attachments/assets/9e839d06-e3c2-4205-b17c-32f44f44f430)

Bug: Text values should be white in dark mode. 

After unlocking the screen and pulling down on the notification drawer, then turning off and on the screen it instantly is corrected:
![2](https://github.com/user-attachments/assets/6abf7665-336a-4ac2-8cd4-d81f28e5abec)

The correct colors are displayed until the next glucose reading comes in 5 minutes later. Then pattern repeats. 

Root cause: Rebuild xDrip with reverting this commit https://github.com/NightscoutFoundation/xDrip/commit/5ac6b761d68a51e6c6b09cc052ebe80b0cb936d5 and issue is gone.

There seems to be bug somewhere with using an older androidx with sdk 24. I think it might be related to this: 
![image](https://github.com/user-attachments/assets/afedcc84-21a7-4061-b1c5-83189acb5f2c)

I can't figure it out and I need help. For now I've compiled the current nightly with minsdk back to 23 and everything is working.

**Comment by Navid200** [comment]
I've never experienced this.  
I use a Pixel 6a.  I wonder why.

**Reply by nickb24** [reply]
Are you running the same version of Android? Do you have 'Dark theme' enabled in the Android display settings?

**Comment by Navid200** [comment]
I have dark theme enabled.  I don't see an option called dark mode.
I am running Android 15 beta.

**Reply by nickb24** [reply]
I think that's why. You are on 15 beta.

**Comment by nickb24** [comment]
I think I'm on to something. Following this idea: https://stackoverflow.com/questions/59910750/notification-with-inverted-color-with-dark-mode

I changed back to 'minSdkVersion 24', but this time changed to 'targetSdkVersion 29'

And now it works.

I've been monitoring logcat for about 10 minutes (2 G6 readings) and everything seems to be working fine.

**Comment by nickb24** [comment]
Small update. Changing targetSDK to 29 breaks the import/export of files to the downloads folder. 

To fix it, in the following file: https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/AndroidManifest.xml

add:  android:requestLegacyExternalStorage="true"

![image](https://github.com/user-attachments/assets/1c222bb0-4429-44fc-81e1-27781c556e1a)

**Comment by nickb24** [comment]
https://github.com/user-attachments/assets/90b0ebe6-ab2d-4fb8-a7e0-e90e45c9b9f5

Well I figured out the issue. It has to do with the specific color you set in the Android theme settings. I had mine set to blue, and with that setting when setting the target sdk to 24 it causes the bug. sdk 23 or 29 don't have it. I was able to reproduce on an emulator as seen in the video. The bug is weird though since as you can see in the video, if you go past the lock screen and pull down the notification once, then it fixes the color. It stays correct until the notification refreshes after 5 minutes (when the next glucose reading comes in).

**Reply by Navid200** [reply]
Are you saying we can tell users to avoid blue and that will be a solution?
Or, are you saying we need to change the xDrip code to allow blue to be used if wanted?

**Comment by nickb24** [comment]
Well I'm not sure if blue is the only color, I didn't try more. But what this tells us is that we confirm there is an Android OS bug when targeting API 24 on a device running Android 14. Can you confirm if Android 15 has the same issue? 

As for what to do for xDrip for now, I guess it's something to keep in mind if other people notice the issue. Changing the API level is definitely something that needs to be incremented higher as Google is making it more difficult for people. But it needs more tests. It's also a good lesson in learning about what happens when you change the target API level. There might be other unnoticed issues.

**Reply by Navid200** [reply]
Android 15 will be released in a month or even less.  We would not be compatible if we didn't do the update that we did.  And we knew that there could be malfunctions.  
We have already addressed two such malfunctions.  And we will deal with this one as well.  

I didn't want people to be negatively surprised when Android 15 is released.
And I am not the only one running Android 15 beta now.  There were requests to update xDrip so that they could use it on Android 15.
We had to update.  We had no choice.  We could have delayed it.  But, all of these detections and corrections would have been delayed also if we did.

So, I choose to look at this as a positive rather than a negative.

**Comment by jamorham** [comment]
I have seen a similar issue on Android 14. It definitely would be good to resolve this.

**Comment by nickb24** [comment]
Great so can we move this over from discussions to the issues list now @Navid200 ?

**Reply by Navid200** [reply]
Yes, I asked you to do that almost 2 weeks ago in the development room:
![Screenshot 2024-08-25 083500](https://github.com/user-attachments/assets/c383043c-3d15-4503-8028-5c0a833bbf26)

**Reply by nickb24** [reply]
Sorry I don't get notifications from Gitter.

**Reply by Navid200** [reply]
Please click (tap) on `Create issue from discussion` in the right pane.
Thanks
